src/backend/auxil/range_alloc.rs

```diff
@@ -0,0 +1,169 @@
+                    // memory fragmentation, so we'll look for it here.
+                    if range.1.end - range.1.start < best_range.end - best_range.start {
+                        *index = range.0;
+                        *best_range = range.1.clone();
+                        // Found a perfect fit, so stop looking.
+                        if range.1.end - range.1.start == length {
+                            break;
+                        }
+                    }
+                },
+                None => { 
+                    best_range_tuple = Some(range.clone());
+                    // Found a perfect fit, so stop looking.
+                    if range.1.end - range.1.start == length {
+                        break;
+                    }
+                }
+            }
+        }
+        best_range_tuple.map(|range| {
+            if range.1.end - range.1.start == length {
+                self.empty_ranges.remove(range.0);
+            } else {
+                self.empty_ranges[range.0].start += length;
+            }
+            range.1.start..(range.1.start + length)
+        })
+    }
+
+    pub fn free_range(&mut self, range: Range<T>) -> bool {
+        assert!(range.start >= self.range.start && range.end <= self.range.end);
+        assert!(range.start <= range.end);
+        if self.empty_ranges.len() == 0 {
+            self.empty_ranges.push(range);
+            return true;
+        }
+        // Input is within range, but before any empty ranges
+        if self.empty_ranges.len() > 0 {
+            if self.empty_ranges[0].start == range.end {
+                self.empty_ranges[0].start = range.start;
+                return true;
+            } else if self.empty_ranges[0].start > range.end {
+                self.empty_ranges.insert(0, range);
+                return true;
+            }
+        }
+        // Input is within range, but after all empty ranges
+        if let Some(last) = self.empty_ranges.last().cloned() {
+            if last.end == range.start {
+                let index = self.empty_ranges.len() - 1;
+                self.empty_ranges[index].end = range.end;
+                return true;
+            } else if last.end < range.start {
+                self.empty_ranges.push(range);
+                return true;
+            }
+        }
+        for i in 0..self.empty_ranges.len() {
+            // Input is immediately to the left of an existing empty range.
+            if range.end == self.empty_ranges[i].start {
+                // Extend this range
+                self.empty_ranges[i].start = range.start;
+                // Merge this into an adjacent range to the left if necessary.
+                if i > 0 && self.empty_ranges[i - 1].end == self.empty_ranges[i].start {
+                    let r = self.empty_ranges.remove(i);
+                    self.empty_ranges[i - 1].end = r.end;
+                    
+                }
+                return true;
```

Comment:
Returning too early, might find inbetween two ranges and we could merge with both neighboring ranges

Reply:
This code already does this.  First it extends one range, then if it's touching its neighbor due to extension then it's merged with its neighbor.

Reply:
Ah right! I got somehow confused by reading 'left' twice